EWS should report test failures from clean-tree to results.webkit.org

JonWBedard opened the pull request:
#### 23a93ba0cddf949d000cadd4198f5f874178a614
<pre>
EWS should report test failures from clean-tree to results.webkit.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=203829">https://bugs.webkit.org/show_bug.cgi?id=203829</a>
rdar://56916405

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/loadConfig.py:
(loadBuilderConfig): Load RESULTS_SERVER_API_KEY from passwords.json.
* Tools/CISupport/ews-build/steps.py:
(RunWebKitTestsWithoutChange.start): Set API key environment variable.
(RunWebKitTestsWithoutChange.setLayoutTestCommand): Upload results to results.webkit.org.
* Tools/CISupport/ews-build/steps_unittest.py:
(TestRunWebKitTestsWithoutChange.test_success):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_success):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_removes_skipped_that_fails):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_fail):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_limit_exceeded):
(TestRunWebKitTestsWithoutChange.test_failure):

Canonical link: <a href="https://commits.webkit.org/257789@main">https://commits.webkit.org/257789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e27a0d578803fc9ac40cda109a5080f5a6161f3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99948 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9113 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33024 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109294 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169531 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86410 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92393 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107197 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105716 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7546 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90827 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34270 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22218 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2914 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23731 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2872 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46107 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/98871 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9005 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43205 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4722 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2748 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3533,6 +3538,17 @@ def commandComplete(self, cmd):
 
     def setLayoutTestCommand(self):
         super(RunWebKitTestsWithoutChange, self).setLayoutTestCommand()
+        if self.getProperty('github.base.ref', DEFAULT_BRANCH) == DEFAULT_BRANCH:
```
Maybe we should also add a condition like: `if CURRENT_HOSTNAME == EWS_BUILD_HOSTNAME` (we don't want to report from uat instance or local testing instances)

xw replied:
Not sure if the check is necessary. Perhaps in the future we'd want to report from the UAT instance to results-dev.safari.apple.com for testing purposes?

aj062 reviewed (commented):
looks good to me

xw commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3510,6 +3511,10 @@ def send_email_for_flaky_failure(self, test_name):
     name = 'run-layout-tests-without-change'
 
+    def start(self):
+        self.workerEnvironment[RESULTS_SERVER_API_KEY] = os.getenv(RESULTS_SERVER_API_KEY)
```
Is it worth logging a warning here if we aren't able to get `RESULTS_SERVER_API_KEY`? Seems like there's a potential to fail silently otherwise.

xw reviewed (approved):
LGTM (though you may want to have someone else r+ this as well, since I'm not an official reviewer)

JonWBedard committed:
[webkit-patch] Parse non-utf characters (Follow-up)
https://bugs.webkit.org/show_bug.cgi?id=245742
<rdar://100466774>

Reviewed by Dewei Zhu.

* Tools/Scripts/webkitpy/common/checkout/diff_parser.py:
(get_diff_converter): Fall back to iso-8859-1 if utf-8 fails.
* Tools/Scripts/webkitpy/style/patchreader.py:
(PatchReader.check): Ditto.

Canonical link: https://commits.webkit.org/257761@main

lingcherd committed:
"Restrict login to IP address" option was removed from Bugzilla's login page. Remove code to set the option to False in bugzilla.py, which is no longer necessary and breaking.
https://bugs.webkit.org/show_bug.cgi?id=249175
<rdar://problem/103269204>

Reviewed by Tim Horton.

* Tools/CISupport/ews-app/ews/common/bugzilla.py:
(BugzillaBeautifulSoup.authenticate):
* Tools/Scripts/webkitpy/common/net/bugzilla/bugzilla.py:
(Bugzilla.authenticate):

Canonical link: https://commits.webkit.org/257762@main

hortont424 committed:
Fix the build after 257681@main
https://bugs.webkit.org/show_bug.cgi?id=249184
rdar://103273163

Unreviewed build fix.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getWindowSceneIdentifierForPaymentPresentation):

Canonical link: https://commits.webkit.org/257763@main

ryanhaddad committed:
[EWS] Add jsc-arm64 queue to PR comment and status bubbles
https://bugs.webkit.org/show_bug.cgi?id=249185
rdar://103273282

Reviewed by Aakash Jain.

* Tools/CISupport/ews-app/ews/common/github.py:
(GitHubEWS):
* Tools/CISupport/ews-app/ews/views/statusbubble.py:
(StatusBubble):

Canonical link: https://commits.webkit.org/257764@main

karlrackler committed:
Fix the build iOS Debug with recent SDK
https://bugs.webkit.org/show_bug.cgi?id=249186
rdar://103273450

Unreviewed build fix.

* Source/WebCore/platform/mediastream/ios/MediaCaptureStatusBarManager.mm:
(-[WebCoreMediaCaptureStatusBarHandler statusBarCoordinator:receivedTapWithContext:completionBlock:]):

Canonical link: https://commits.webkit.org/257765@main

ryanhaddad committed:
[Gardening] Skip two more fast/forms/ios/file-upload-panel* tests
https://bugs.webkit.org/show_bug.cgi?id=249193
rdar://103275061

Unreviewed test gardening.

* LayoutTests/platform/ios/TestExpectations: Also, unskip
a test that was wrongfully blamed for the crash.

Canonical link: https://commits.webkit.org/257766@main

JonWBedard committed:
[ews-build.webkit.org] Use new-style shell commands for services testing
https://bugs.webkit.org/show_bug.cgi?id=249029
rdar://103186946

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(RunWebKitPerlTests): Make step ShellCommandNewStyle.
(RunBuildWebKitOrgUnitTests): Ditto.
(RunEWSUnitTests): Ditto.
(RunResultsdbpyTests): Ditto.
(WebKitPyTest): Ditto.
(WebKitPyTest.run): Renmaed from WebKitPyTest.start.
(RunBuildWebKitOrgUnitTests.start): Deleted.
(WebKitPyTest.start): Renamed to WebKitPyTest.run.

Canonical link: https://commits.webkit.org/257767@main

rniwa committed:
Use dump-as-markup.js in fast/attachment/attachment-select-on-click.html
https://bugs.webkit.org/show_bug.cgi?id=249121

Reviewed by Darin Adler.

This test is testing selection across attachment element. Use dump-as-markup.js for clarity.

* LayoutTests/fast/attachment/attachment-select-on-click-expected.txt: Added.
* LayoutTests/fast/attachment/attachment-select-on-click.html:
* LayoutTests/platform/gtk/fast/attachment/attachment-select-on-click-expected.txt: Removed.
* LayoutTests/platform/ios/fast/attachment/attachment-select-on-click-expected.txt: Removed.
* LayoutTests/platform/mac/fast/attachment/attachment-select-on-click-expected.png: Removed.
* LayoutTests/platform/mac/fast/attachment/attachment-select-on-click-expected.txt: Removed.
* LayoutTests/platform/wincairo/fast/attachment/attachment-select-on-click-expected.txt: Removed.
* LayoutTests/platform/wpe/fast/attachment/attachment-select-on-click-expected.txt: Removed.

Canonical link: https://commits.webkit.org/257768@main

mattwoodrow committed:
Add TransferImageBuffer message
https://bugs.webkit.org/show_bug.cgi?id=249111
<rdar://problem/103232081>

Reviewed by Kimmo Kinnunen.

This implements transfering the backend of a RemoteImageBuffer from its existing RemoteRenderingBackend/WorkQueue to its new RRB.
This will likely be replaced in the future with the new way of sharing resources in the GPUP.

When we create a new ImageBuffer in the WebProcess, and want to transfer contents, we send the message across the connection for the
new buffer.
The GPUP side dispatches messages to the main thread to find the RemoteRenderingBackend, dispatches to the existing WorkQueue to take
ownership of the old buffer and dispatches the result back to the new WorkQueue to complete the transfer.

RemoteImageBuffer sends the same didCreateImageBufferBackend message back once the buffer is transferred (as happens when a remote buffer
is allocated), so that WebProcess attempts to block and wait for the buffer to be available succeed.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::transferToNewContext):
(WebCore::ImageBufferBackend::parameters):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::transferToNewContext):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
(WebKit::GPUConnectionToWebProcess::findRemoteRenderingBackend):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::~RemoteImageBuffer):
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::createTransfer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::transferImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::resourceOwner const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheImageBuffer):
(WebKit::RemoteResourceCache::cachedImageBuffer const):
(WebKit::RemoteResourceCache::takeImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::sendToStreamWithAsyncReply):

Canonical link: https://commits.webkit.org/257769@main

mwyrzykowski committed:
[WebGPU] Add support for GPU Frame capture
https://bugs.webkit.org/show_bug.cgi?id=249011
<radar://103045608>

Reviewed by Dean Jackson.

Add support for capturing the nth-frame during rendering
via notifyutil -p com.apple.WebKit.WebGPU.CaptureFrame as
well as the initial frame which includes all of the resource
creation commands in the gputrace via ...ToggleCaptureFirstFrame

Requires launching the process with __XPC_METAL_CAPTURE_ENABLED=1

These traces end up in /var/folders/.../com.apple.WebKit.GPU+org.webkit.MiniBrowser/

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::Device::createCommandEncoder):
Capture the frame starting before the command encoder is created.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::frameCaptureEnabled):
(WebGPU::captureFirstFrame):
(WebGPU::registerForFrameCapture):
(WebGPU::Device::Device):
(WebGPU::Device::captureFrameIfNeeded const):
Add support MTLCaptureManager.

* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commitMTLCommandBuffer):
Stop frame capture after queue is committed.

Canonical link: https://commits.webkit.org/257770@main

rniwa committed:
[Live Range Selection] fast/events/autoscroll-in-textarea.html fails
https://bugs.webkit.org/show_bug.cgi?id=249188

Reviewed by Darin Adler.

The test failure was caused by getSelection().toString() no longer returning
the selected text inside textarea. Fixed it by using textarea's selectionStart
and selectionEnd to get the selected substring.

* LayoutTests/fast/events/autoscroll-in-textarea.html:

Canonical link: https://commits.webkit.org/257771@main

nt1m committed:
Remove font-palette from font shorthand
https://bugs.webkit.org/show_bug.cgi?id=248949
rdar://103119662

Reviewed by Cameron McCormack.

It was resolved in https://github.com/w3c/csswg-drafts/issues/7832 to remove font-palette from the font shorthand.

WPT synced from Mozilla's change: https://phabricator.services.mozilla.com/D164128

* LayoutTests/fast/css/font-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-vs-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-vs-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention.html:
* LayoutTests/platform/win/fast/css/font-shorthand-expected.txt:
* LayoutTests/platform/win/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontShorthandValue):
* Source/WebCore/css/StylePropertyShorthand.h:

Canonical link: https://commits.webkit.org/257772@main

aj062 committed:
[ews] Commit-Queue pre-maturely finished the build after re-run-layout-tests
https://bugs.webkit.org/show_bug.cgi?id=249049
rdar://103197446

Reviewed by Ryan Haddad.

* Tools/CISupport/ews-build/steps.py:
(ReRunWebKitTests.evaluateCommand):

Canonical link: https://commits.webkit.org/257773@main

rniwa committed:
Use dump-as-markup.js in fast/dom/delete-contents.html
https://bugs.webkit.org/show_bug.cgi?id=249172

Reviewed by Wenson Hsieh.

Use dump-as-markup.js in this test for clarity.

* LayoutTests/fast/dom/delete-contents-expected.txt:
* LayoutTests/fast/dom/delete-contents.html:

Canonical link: https://commits.webkit.org/257774@main

DimStar77 committed:
Fix build with Ruby 3.2
https://bugs.webkit.org/show_bug.cgi?id=246743

Reviewed by Ross Kirsling.

File.exists has been declared deprecated since Ruby 2.1 (2013) and with Ruby 3.2 gets removed for good.

* Source/JavaScriptCore/offlineasm/config.rb:
* Source/JavaScriptCore/offlineasm/parser.rb:
* Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb:

Canonical link: https://commits.webkit.org/257775@main

grorg committed:
P3 colorspace support for WebGL canvas on iOS
https://bugs.webkit.org/show_bug.cgi?id=249180
rdar://101663634

Reviewed by Simon Fraser.

Remove the flag restricting this to macOS.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.idl:

Canonical link: https://commits.webkit.org/257776@main

kshukuwa committed:
[Curl] Remove unnecessary function from ResourceResponseCurl
https://bugs.webkit.org/show_bug.cgi?id=249195

Reviewed by Fujii Hironori.

Remove ResourceResponse::setCertificateInfo. Also, we configure
CertificateInfo in a ResourceResponse constructor.

* Source/WebCore/platform/network/curl/CurlResourceHandleDelegate.cpp:
(WebCore::CurlResourceHandleDelegate::curlDidReceiveResponse):
* Source/WebCore/platform/network/curl/ResourceResponse.h:
* Source/WebCore/platform/network/curl/ResourceResponseCurl.cpp:
(WebCore::ResourceResponse::ResourceResponse):
(WebCore::ResourceResponse::setCertificateInfo): Deleted.
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::curlDidReceiveResponse):

Canonical link: https://commits.webkit.org/257777@main

beidson committed:
Add some notifications related WKWebsiteDataStoreDelegate calls
https://bugs.webkit.org/show_bug.cgi?id=249164
rdar://103271182

Reviewed by Ben Nham.

Adds variants of the permissions and showNotification calls to _WKWebSiteDataStoreDelegate

Also fixes a few other issues in the push/notifications code paths.
And adds some logging.

* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::processPushMessage):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::getPendingPushMessages):
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
(WebKit::Daemon::Coder<WTF::UUID>::encode): Deleted.
(WebKit::Daemon::Coder<WTF::UUID>::decode): Deleted.
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder<WTF::UUID>::encode):
(WebKit::Daemon::Coder<WTF::UUID>::decode):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
(-[_WKNotificationData initWithCoreData:dataStore:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationDataInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::showNotification):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::showServiceWorkerNotification):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::showNotification):
(WebKit::WebsiteDataStoreClient::notificationPermissions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
(messageDictionary):
(-[FirePushEventDataStoreDelegate notificationPermissionsForWebsiteDataStore:]):
(-[FirePushEventDataStoreDelegate websiteDataStore:showNotification:]):

Canonical link: https://commits.webkit.org/257778@main

graouts committed:
perspective calc() values aren't clipped to 0
https://bugs.webkit.org/show_bug.cgi?id=249151

Reviewed by Simon Fraser and Dean Jackson.

The "perspective" property is specified to only allow non-negative values,
see https://www.w3.org/TR/css-transforms-2/#perspective-property. However,
we don't clip values that are computed to less than 0, for instance using a
calc() expression. Instead, we deem them invalid internally and thus consider
as if "none" was provided, which has implications when blending. Indeed, when
interpolating between a valid value and "none", we use discrete interpolation.

We now correctly clip values below 0.

Note that the relevant WPT test still fails, because the test itself is not
valid. It is being addressed in a dedicated WPT PR and will be imported back
into WebKit yielding a PASS result. See https://github.com/web-platform-tests/wpt/pull/37457.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/perspective-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/perspective-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumePerspective):

Canonical link: https://commits.webkit.org/257779@main

rniwa committed:
Use dump-as-markup.js in fast/dom/delete-contents.html
https://bugs.webkit.org/show_bug.cgi?id=249124

Reviewed by Darin Adler.

Use dump-as-markup.js in this test for clarity.

* LayoutTests/fast/dom/delete-contents-expected.txt:
* LayoutTests/fast/dom/delete-contents.html:

Canonical link: https://commits.webkit.org/257780@main

rniwa committed:
[Live Range Selection] fast/text/pre-wrap-trailing-tab.html fails
https://bugs.webkit.org/show_bug.cgi?id=249182

Reviewed by Darin Adler.

The test failure was caused by selection no longer getting canonicalized.
Explicitly canonicalize the selection ends by extending forward then backward.

* LayoutTests/fast/text/pre-wrap-trailing-tab.html:

Canonical link: https://commits.webkit.org/257781@main

philn committed:
[GStreamer] Support for creating VideoFrame from current video element frame
https://bugs.webkit.org/show_bug.cgi?id=248954

Reviewed by Xabier Rodriguez-Calvar.

The videoFrameForCurrentTime() method is needed for eg, feeding a <video> element frame to a
WebCodec video encoder.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::videoFrameForCurrentTime):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/257782@main

carlosgcampos committed:
[GTK][WPE] Add webkit_web_hit_test_result_get_js_node() and deprecate the rest of the DOM APIs
https://bugs.webkit.org/show_bug.cgi?id=249020

Reviewed by Michael Catanzaro.

* Source/WebKit/WebProcess/InjectedBundle/API/glib/DOM/WebKitDOMNode.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/DOM/WebKitDOMObject.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
(webkit_frame_get_js_value_for_dom_object):
(webkit_frame_get_js_value_for_dom_object_in_script_world):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp:
(webkitWebHitTestResultGetProperty):
(webkitWebHitTestResultSetProperty):
(webkit_web_hit_test_result_class_init):
(webkitWebHitTestResultCreate):
(webkit_web_hit_test_result_get_node):
(webkit_web_hit_test_result_get_js_node):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMNode.h:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMObject.h:
* Source/WebKit/WebProcess/InjectedBundle/API/wpe/DOM/WebKitDOMNode.h:
* Source/WebKit/WebProcess/InjectedBundle/API/wpe/DOM/WebKitDOMObject.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/FrameTest.cpp:
(WebKitFrameTest::testJavaScriptValues):
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(contextMenuCallback):

Canonical link: https://commits.webkit.org/257783@main

csaavedra committed:
Debug build fixes with assertions disabled
https://bugs.webkit.org/show_bug.cgi?id=249129

Reviewed by Philippe Normand.

When disabling assertions in a Debug build there are
plenty of unused variables that break the build due
to -Werror. Most of those are due to using NDEBUG
when ASSERT_ENABLED is needed. Fix those.

* Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
* Source/WTF/wtf/FastMalloc.cpp:
* Source/WebCore/Modules/indexeddb/server/MemoryIDBBackingStore.cpp:
(WebCore::IDBServer::MemoryIDBBackingStore::clearObjectStore):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::validateSDP const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::notificationPostTimerFired):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::AccessibilityRenderObject):
(WebCore::AccessibilityRenderObject::detachRemoteParts):
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::removedLastRef):
* Source/WebCore/layout/formattingContexts/FormattingContext.cpp:
(WebCore::Layout::FormattingContext::validateGeometryConstraintsAfterLayout const):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::~CachedResource):
* Source/WebCore/platform/PODRedBlackTree.h:
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::CDMFactoryClearKey::createCDM):
(WebCore::CDMInstanceSessionClearKey::loadSession):
(WebCore::CDMInstanceSessionClearKey::removeSessionData):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::codePointSupportIndex):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::handleErrorConditionFromStreamingThread):
* Source/WebCore/rendering/RenderFragmentedFlow.cpp:
(WebCore::RenderFragmentedFlow::removeRenderBoxFragmentInfo):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::clippedOverflowRect const):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::paintObject):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::checkLayoutAttributesConsistency):

Canonical link: https://commits.webkit.org/257784@main

whsieh committed:
[iOS] Scrolling after element focus does not avoid input views when OOP keyboard is enabled
https://bugs.webkit.org/show_bug.cgi?id=249159
rdar://88057475

Reviewed by Aditya Keerthi.

Scrolling to reveal the focused element is currently broken when OOP keyboard is enabled. This is
because OOP keyboard causes `UIKeyboardWillShowNotification` to be delivered to the app
asynchronously (after communicating with the input UI process), rather than dispatched immediately
under the call to `-reloadInputViews`. WebKit currently depends on the latter, since we expect the
keyboard's final geometry (after animation) to be set by the time the post-layout editor state
arrives in the UI process and we reveal the focused element using `-_zoomToRevealFocusedElement`.
However, since this geometry is sent through the "KeyboardWillShow" notification which (normally)
arrives after the editor state, we end up proceeding with `-_zoomToRevealFocusedElement` without
knowing the final keyboard geometry, so we end up only scrolling to keep the field centered within
the web view (ignoring the obscuring bounds of the software keyboard). In some cases (e.g. when
focusing an input field in some system views), this can cause focused input fields to be completely
obscured by the keyboard.

This is somewhat tricky to fix, since it's not guaranteed that the "KeyboardWillShow" notification
will always arrive after the editor state or vice versa. To complicate things further, in the Mail
compose case (i.e. when the web view's scroller has `-firstResponderKeyboardAvoidanceEnabled` set to
`NO`), we'll actually zoom to reveal the focused element only after `UIKeyboardDidShowNotification`
instead, since the keyboard geometry isn't available before the animation has finished (see
https://commits.webkit.org/246159@main for more context). And of course, this fix also needs to
continue working with OOP keyboard disabled, where the "KeyboardWillShow" notification is dispatched
before we've even begun waiting for the next editor state.

To handle all of the above scenarios, we add a helper class that encapsulates logic for deferring
calls to `-_zoomToRevealFocusedElement` after an element is focused. This helper class,
`WebKit::RevealFocusedElementDeferrer`, maintains a set of flags that enumerate the three distinct
reasons why we could be deferring this call: "waiting for an editor state", "waiting for a
`KeyboardWillShow` notification", and "waiting for a `KeyboardDidShow` notification". As
`WKContentView` receives keyboard notification from UIKit or receives an editor state update from
the web process, it removes the appropriate deferral reasons from this helper object. Finally, when
there are no longer any deferral reasons, we call into `-_zoomToRevealFocusedElement` and clear out
the helper.

Test: fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html

* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard-expected.txt: Added.
* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html: Added.

Add a new layout test to exercise this scenario. All existing layout tests are actually insufficient
to cover the changes in this fix, since a test needs to:

-   Show the software keyboard (by default, layout tests run with the hardware keyboard connected).
-   Check that we scrolled the focused input into (roughly) the center of the visible viewport, not
    just onto the screen.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.h: Added.
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.mm: Added.
(WebKit::RevealFocusedElementDeferrer::RevealFocusedElementDeferrer):
(WebKit::RevealFocusedElementDeferrer::create):
(WebKit::RevealFocusedElementDeferrer::fulfill):

Add the new helper class. We protect the deferrer here since it expects its client (`WKContentView`)
to clear any references to itself when `-_zoomToRevealFocusedElement` is invoked.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _keyboardWillShow]):
(-[WKContentView _keyboardDidShow]):
(-[WKContentView _zoomToRevealFocusedElement]):
(-[WKContentView resetShouldZoomToFocusRectAfterShowingKeyboard]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

Either call `-_zoomToRevealFocusedElement` immediately, or create a `RevealFocusedElementDeferrer`,
passing in deferral reasons as appropriate. See comments above for more details.

(-[WKContentView _elementDidBlur]):
(-[WKContentView _didUpdateEditorState]):
(-[WKContentView _zoomToFocusRectAfterShowingKeyboardIfNeeded]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/257785@main

anttijk committed:
[@property] Parse identifier list syntax
https://bugs.webkit.org/show_bug.cgi?id=249161
<rdar://problem/103262360>

Reviewed by Alan Baradlay.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* Source/WebCore/css/parser/CSSPropertySyntax.cpp:
(WebCore::CSSPropertySyntax::parseComponent):

Allow ident+ and ident# syntax.

(WebCore::CSSPropertySyntax::parse):
(WebCore::CSSPropertySyntax::typeForTypeName):

Factor into a function.

* Source/WebCore/css/parser/CSSPropertySyntax.h:

Canonical link: https://commits.webkit.org/257786@main

philn committed:
[WPE][GTK] Enable permissions request API
https://bugs.webkit.org/show_bug.cgi?id=247065

Reviewed by Carlos Garcia Campos.

This is exposed in the GLib API as a WebKitPermissionStateQuery signalled by the WebView using the
`query-permission-state` GObject signal. When the application has a handler for this signal it has
to eventually call `webkit_permission_state_query_finish()` with a valid `WebKitPermissionState`. If
not handled, the default result is `WEBKIT_PERMISSION_STATE_PROMPT`, thus instructing the User-Agent
that it has to explicitely present a prompt asking for permission to the given feature.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitAutocleanups.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.cpp: Added.
(_WebKitPermissionStateQuery::_WebKitPermissionStateQuery):
(_WebKitPermissionStateQuery::~_WebKitPermissionStateQuery):
(webkitPermissionStateQueryCreate):
(webkit_permission_state_query_ref):
(webkit_permission_state_query_unref):
(webkit_permission_state_query_get_name):
(webkit_permission_state_query_get_security_origin):
(webkit_permission_state_query_finish):
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQueryPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(webkitWebViewPermissionStateQuery):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/webkit.h:
* Source/WebKit/UIProcess/API/gtk/webkit2.h:
* Source/WebKit/UIProcess/API/wpe/webkit.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewQueryPermissionRequests):
(beforeAll):

Canonical link: https://commits.webkit.org/257787@main

philn committed:
[GStreamer] Handle latency messages
https://bugs.webkit.org/show_bug.cgi?id=248813

Reviewed by Xabier Rodriguez-Calvar.

When a latency message is received on the bus the application should recalculate and distribute a
new latency. Also included in this patch is a drive-by improvement in state-change message handling
in the audio file reader, where we process only the notification from the top-level pipeline.

* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
(WebCore::AudioDestinationGStreamer::handleMessage):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::AudioFileReader::handleMessage):
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp:
(WebCore::AudioSourceProviderGStreamer::AudioSourceProviderGStreamer):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::connectSimpleBusMessageCallback):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::InnerDecoder::handleMessage):

Canonical link: https://commits.webkit.org/257788@main

JonWBedard committed:
EWS should report test failures from clean-tree to results.webkit.org
https://bugs.webkit.org/show_bug.cgi?id=203829
rdar://56916405

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/loadConfig.py:
(loadBuilderConfig): Load RESULTS_SERVER_API_KEY from passwords.json.
* Tools/CISupport/ews-build/steps.py:
(RunWebKitTestsWithoutChange.start): Set API key environment variable.
(RunWebKitTestsWithoutChange.setLayoutTestCommand): Upload results to results.webkit.org.
* Tools/CISupport/ews-build/steps_unittest.py:
(TestRunWebKitTestsWithoutChange.test_success):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_success):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_removes_skipped_that_fails):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_fail):
(TestRunWebKitTestsWithoutChange.test_run_subtest_tests_limit_exceeded):
(TestRunWebKitTestsWithoutChange.test_failure):

Canonical link: https://commits.webkit.org/257789@main
